### PR TITLE
fix(ivy): avoid remote scoping if it's not actually required

### PR DESCRIPTION
**Pull request by alxhub**
Currently, ngtsc decides to use remote scoping if the compilation of a
component may create a cyclic import. This happens if there are two
components in a scope (say, A and B) and A directly uses B. During
compilation of B ngtsc will then note that if B were to use A, a cycle would
be generated, and so it will opt to use remote scoping for B.

ngtsc already uses the R3TargetBinder to correctly track the imports that
are actually required, for future cycle tracking. This commit expands that
usage to not trigger remote scoping unless B actually does consume A in its
template.

### CODE REVIEW

File: packages/compiler-cli/test/ngtsc/ngtsc_spec.ts [hunk 1]
```diff
@@ -3934,6 +3967,57 @@ export const Foo = Foo__PRE_R3__;
       expect(jsContents).toContain('styles: ["h1[_ngcontent-%COMP%] {font-size: larger}"]');
     });
   });
+
+  xit('should not suck', () => {
```
Comment (JoostK) [inline packages/compiler-cli/test/ngtsc/ngtsc_spec.ts]: ROFL 🤣

Comment (alxhub) [reply packages/compiler-cli/test/ngtsc/ngtsc_spec.ts]: Hahaha. That's my usual test description for reproducing bugs within the compiler's test suite.

File: packages/compiler-cli/src/ngtsc/annotations/src/component.ts [hunk 2]
```diff
@@ -322,52 +322,87 @@ export class ComponentDecoratorHandler implements
...
-      const cycleDetected = directives.some(dir => this._isCyclicImport(dir.expression, context)) ||
-          Array.from(pipes.values()).some(pipe => this._isCyclicImport(pipe, context));
+      // Next, the component template AST is bound using the R3TargetBinder. This produces an
+      // BoundTarget, which is similar to a ts.TypeChecker.
+      const binder = new R3TargetBinder(matcher);
+      const bound = binder.bind({template: metadata.template.nodes});
+
+      // The BoundTarget knows which directives and pipes matched the template.
+      const usedDirectives = bound.getUsedDirectives();
+      const usedPipes = bound.getUsedPipes().map(name => pipes.get(name) !);
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/annotations/src/component.ts]: This non-null assertion operator is actually a problem, I think. `T3TargetBinder`, or more specifically `TemplateBinder`:

https://github.com/angular/angular/blob/5ad2097be857dd037706a83c44016346237f4d59/packages/compiler/src/render3/view/t2_binder.ts#L444-L447

tracks pipes just by name regardless of their definition. So, a missing pipe results in a miss here.

File: packages/compiler-cli/src/ngtsc/annotations/src/component.ts [hunk 3]
```diff
@@ -322,52 +322,87 @@ export class ComponentDecoratorHandler implements
...
       const pipes = new Map<string, Expression>();
       for (const pipe of scope.compilation.pipes) {
         pipes.set(pipe.name, this.refEmitter.emit(pipe.ref, context));
       }
 
-      // Scan through the references of the `scope.directives` array and check whether
-      // any import which needs to be generated for the directive would create a cycle.
-      const cycleDetected = directives.some(dir => this._isCyclicImport(dir.expression, context)) ||
-          Array.from(pipes.values()).some(pipe => this._isCyclicImport(pipe, context));
+      // Next, the component template AST is bound using the R3TargetBinder. This produces an
```
Comment (gkalpak) [inline packages/compiler-cli/src/ngtsc/annotations/src/component.ts]: _an_ --> _a_ :v:

File: packages/compiler-cli/src/ngtsc/annotations/src/component.ts (hunk 2 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/component.ts]: You're absolutely right. I'll file an issue to clean that up.

**Comment by alxhub** [comment]
I am the approver for compiler-cli.

**Pull request closed without merging**